mage/query-modules/python/gnn-link-prediction.md

```diff
@@ -0,0 +1,375 @@
+- support for both **homogeneous** and **heterogeneous** graphs
+- support for **disconnected** graphs
+- its applicability to use it as a **recommendation system**
+- a **semi-inductive** link prediction setup where a larger, updated graph is used for the **inference**
+- an **inductive** link prediction setup in which **training** and **inference** graphs are different
+- **transductive** graph splitting(training and validation sets)
+- **Graph attention layer** aggregates information using an attention mechanism from the first-hop neighbourhood. Introduced by [Velickovic et al.](https://arxiv.org/pdf/1710.10903.pdf)
+- **GraphSAGE layer** extends the usability of graph neural networks to large-scale graphs. Introduced by [Hamilton et al.](https://arxiv.org/pdf/1706.02216.pdf)
+- **mlp** and **dot** predictors are used for combining node scores to edge scores
+- **ADAM** and **SGD** optimizers are used for training neural networks
+- support for **batch training**
+- **parallel graph sampling**  is done using multiple threads
+- **negative graph sampling** is a sampling where the final graph consists only of edges that don't exist
+- evaluating the model's **training performance** using a variety of metrics like **AUC, Precision, Recall, Accuracy, Confusion matrix**
+- evaluating the model's **recommendation performance** with **Precision@k, Recall@k, F1@k and Average Precision** metrics
+
+If you want to try-out our implementation, head to **[github/memgraph/mage](https://github.com/memgraph/mage)** and find `python/link_prediction.py`. Feel free to give us some :star: if you like the code. The easiest way to test `link-prediction` is by downloading [Memgraph Platform](https://memgraph.com/download) and using some of the preloaded datasets in **Memgraph Lab**.
+
+There are some things you should be careful about when using `link prediction`:
+- features of all nodes should be called the same(e.g saved as **'features'** property in **Memgraph**) 
+- model's performance on the validation set is obtained using **transductive** splitting mode, while **inductive** dataset split is not yet supported. You can find more information about graph splitting on slides of [Graph Machine Learning course](http://web.stanford.edu/class/cs224w/slides/08-GNN-application.pdf) offered by **Stanford**. 
+- to improve performance, `self-loop` is added to each node with the edge-type set to `self`
+- also, the user can set the flag to automatically add **reverse edges** to each existing edge and hence, convert a **directed** graph to a **bidirected** one. If the source and destination nodes of the edge are the same, **reverse edge type** will be the same as the original **edge type**. Otherwise, the prefix **rev_** will be added to the original **edge type**. See the implementation part to further see why are **self-loops** and **reverse edges** very important in ML training and how you can get into problems if your graph is already **undirected** :thinking_face: 
```

Comment:
```suggestion
- the user can set the flag to automatically add **reverse edges** to each existing edge and hence, convert a **directed** graph to a **bidirected** one. If the source and destination nodes of the edge are the same, **reverse edge type** will be the same as the original **edge type**. Otherwise, the prefix **rev_** will be added to the original **edge type**. See the implementation part to further see why are **self-loops** and **reverse edges** very important in ML training and how you can get into problems if your graph is already **undirected** :thinking_face: 
```